### PR TITLE
MANTA-4720 Update buckets-api and buckets-mdplacement to new node version

### PR DESCRIPTION
**Pull request by bahamas10**
From a high level this change:

1. Updates node from v0.10 -> v6
2. Bumps dependencies to work with the new node

On a more specific level:

1. upgrade node v0.10.48 -> v6.17.0
2. upgrade dep bignum v0.6.0 -> v0.13.0
3. upgrade dep dtrace-provider v0.2.8 -> v0.8.8
4. upgrade dep mime v1.2.11 -> v1.4.0
5. upgrade dep sshpk 1.13.1 -> 1.16.1
6. upgrade dep restify 2.6.3 -> 6.4.0
7. remove deps libuuid and node-uuid
8. add dep uuid
9. add deps restify-clients and restify-errors (used to be in restify v2)
10. swap `req._time` with `req.time()` (supported restify interface)
11. swap `res.headers()` with `res.getHeaders()` (supported restify interface)
12. swap `req.params` with `req.query` where applicable (query no longer rolled into params in restify v6, this was already done in `auth.js`)
13. remove deprecated domain logic for restify / http request
14. fix RequestCaptureStream to stderr (fallout from syslog removal)

I tested this manually with the `manta` bash function that wraps `curl`, as well as the `mbucket` command.  Also, this commit was tested with `node-manta` on the `buckets` branch with the 2 bucket related tests:

Tested on commit 11997259d06aba991e486698dabe66d6c13e1537

```
dave - temple linux ~/dev/buckets/node-manta (git:buckets) $ node test/integration/buckets-client-basic.test.js 
TAP version 13
# Subtest: buckets client basic
    1..0
ok 1 - buckets client basic # time=35.058ms

# Subtest: setup client
    1..0
ok 2 - setup client # time=13.158ms

# Subtest: isBucketsSupported
    ok 1 - should not error
    ok 2 - isSupported
    1..2
ok 3 - isBucketsSupported # time=645.292ms

# Subtest: createBucket
    ok 1 - should not error
    1..1
ok 4 - createBucket # time=251.884ms

# Subtest: headBucket
    ok 1 - should not error
    1..1
ok 5 - headBucket # time=341.877ms

# Subtest: createListBucketsStream
    ok 1 - got at least one bucket in listing, theBuckets names: bar, bucket-1, bucket-2, bucket-3, dave-1, dave-2, dave-3, foo, manta-4720, mb-bucket, node-manta-test-buckets-client-basic-a7b04404-bucket, node-manta-test-buckets-client-basic-b52f3841-bucket, slashes, test-bucket, trent
    1..1
ok 6 - createListBucketsStream # time=445.778ms

# Subtest: createBucketObject
    ok 1 - should not error
    1..1
ok 7 - createBucketObject # time=693.779ms

# Subtest: headBucketObject
    ok 1 - should not error
    ok 2 - expect truthy value
    ok 3 - should be equal
    ok 4 - should be equal
    ok 5 - should be equal
    1..5
ok 8 - headBucketObject # time=308.71ms

# Subtest: getBucketObject
    ok 1 - should not error
    ok 2 - expect truthy value
    ok 3 - should be equal
    ok 4 - should be equal
    ok 5 - should be equal
    ok 6 - expect truthy value
    ok 7 - should be equivalent strictly
    1..7
ok 9 - getBucketObject # time=1433.011ms

# Subtest: putBucketObjectMetadata
    ok 1 - should not error
    ok 2 - expect truthy value
    ok 3 - should be equal
    1..3
ok 10 - putBucketObjectMetadata # time=509.908ms

# Subtest: createListBucketObjectsStream
    ok 1 - got one object in bucket node-manta-test-buckets-client-basic-b52f3841-bucket
    ok 2 - should be equal
    1..2
ok 11 - createListBucketObjectsStream # time=513.587ms

# Subtest: deleteBucketObject
    ok 1 - should not error
    1..1
ok 12 - deleteBucketObject # time=508.708ms

# Subtest: deleteBucket
    ok 1 - should not error
    1..1
ok 13 - deleteBucket # time=838.399ms

# Subtest: have we tested all the client methods once?
    ok 1 - the set of client methods remaining to test is empty: []
    1..1
ok 14 - have we tested all the client methods once? # time=2.318ms

# Subtest: teardown
    1..0
ok 15 - teardown # time=2.031ms

1..15
# time=6561.562ms
dave - temple linux ~/dev/buckets/node-manta (git:buckets) $ node test/integration/buckets-client-condreq.test.js 
TAP version 13
# Subtest: buckets client conditional requests
    1..0
ok 1 - buckets client conditional requests # time=19.078ms

# Subtest: setup: client
    1..0
ok 2 - setup: client # time=13.568ms

# Subtest: setup: create bucket: node-manta-test-buckets-client-condreq-8500e03b-bucket
    ok 1 - should not error
    1..1
ok 3 - setup: create bucket: node-manta-test-buckets-client-condreq-8500e03b-bucket # time=772.424ms

# Subtest: setup: create object: node-manta-test-buckets-client-condreq-8500e03b-object
    ok 1 - should not error
    ok 2 - expect truthy value
    1..2
ok 4 - setup: create object: node-manta-test-buckets-client-condreq-8500e03b-object # time=607.673ms

# Subtest: HeadBucketObject: if-modified-since
    ok 1 - should not error
    ok 2 - expect truthy value
    ok 3 - should be equal
    ok 4 - should be equal
    1..4
ok 5 - HeadBucketObject: if-modified-since # time=415.93ms

# Subtest: teardown: delete object
    ok 1 - should not error
    1..1
ok 6 - teardown: delete object # time=504.473ms

# Subtest: teardown: delete bucket
    ok 1 - should not error
    1..1
ok 7 - teardown: delete bucket # time=1125.106ms

# Subtest: teardown: client
    1..0
ok 8 - teardown: client # time=0.83ms

1..8
# time=3469.311ms
```

### CODE REVIEW

**Comment by trentm** [comment]
> 12. swap req.params with req.query where applicable (query no longer rolled into params in restify v6, this was already done in auth.js)

That was a change in restify.plugins.queryParser() default behaviour in v4, v5, BTW. Gory details in https://jira.joyent.us/browse/ZAPI-744
Our usage without options to it should be fine. If using restify v4, one has to watch out.

File: package.json [hunk 1]
```diff
@@ -15,7 +15,7 @@
         "artedi": "1.1.1",
         "assert-plus": "0.1.5",
         "backoff": "2.3.0",
-        "bignum": "0.6.0",
+        "bignum": "0.13.0",
```
Comment (trentm) [inline package.json:18]: "bunyan": "0.22.1",

Really should update bunyan to latest 1.x: ^1.8.12

File: package.json [hunk 2]
```diff
@@ -24,22 +24,24 @@
...
         "mahi": "2.3.1",
-        "mime": "1.2.11",
+        "mime": "1.4.0",
         "moray": "3.4.1",
         "once": "1.3.0",
-        "restify": "2.6.3",
-        "sshpk": "1.13.1",
+        "restify": "6.4.0",
+        "restify-clients": "2.6.7",
+        "restify-errors": "7.0.0",
```
Comment (trentm) [inline package.json]: Hopefully not, but there is potential fallout from this new restify-errors.
Note that restify v6 uses restify-errors v5. Why using v7 here? It might be
fine.

File: package.json [hunk 3]
```diff
@@ -24,22 +24,24 @@
...
+        "sshpk": "1.16.1",
         "tab": "1.0.0",
         "vasync": "^2.2.0",
         "verror": "^1.9.0",
```
Comment (trentm) [inline package.json:46]: Would be nice to bump to verror 1.10.0 to get
https://github.com/joyent/node-verror/issues/49
Though not required, of course.

File: package.json [hunk 4]
```diff
@@ -24,22 +24,24 @@
...
         "deep-equal": "0.0.0",
-        "dtrace-provider": "0.2.8",
+        "dtrace-provider": "0.8.8",
         "fast": "2.8.1",
         "http-signature": "1.1.0",
         "jsprim": "^1.3.1",
         "kang": "1.1.0",
         "keep-alive-agent": "0.0.1",
         "keyapi": "git+https://github.com/joyent/keyapi.git#e14b3d58",
         "libmanta": "git+https://github.com/joyent/node-libmanta.git#v1.3.0",
```
Comment (trentm) [inline package.json:33]: There are a number of `gyp ERR! build error` for, I'm guessing,
optionalDependencies in the build lot. E.g. for old bunyan:


```
...
   ^
../dtrace_provider.cc: In static member function 'static v8::Handle<v8::Value> node::DTraceProvider::Fire(const int&)':
../dtrace_provider.cc:186:3: warning: control reaches end of non-void function [-Wreturn-type]
   }
   ^
DTraceProviderBindings.target.mk:99: recipe for target 'Release/obj.target/DTraceProviderBindings/dtrace_provider.o' failed
make[1]: *** [Release/obj.target/DTraceProviderBindings/dtrace_provider.o] Error 1
make[1]: Leaving directory '/root/joy/manta-buckets-api/node_modules/bunyan/node_modules/dtrace-provider/build'
gyp ERR! build error
gyp ERR! stack Error: `make` failed with exit code: 2
gyp ERR! stack     at ChildProcess.onExit (/root/joy/manta-buckets-api/build/node/lib/node_modules/npm/node_modules/node-gyp/lib/build.js:276:23)
gyp ERR! stack     at emitTwo (events.js:106:13)
gyp ERR! stack     at ChildProcess.emit (events.js:191:7)
gyp ERR! stack     at Process.ChildProcess._handle.onexit (internal/child_process.js:219:12)
gyp ERR! System SunOS 5.11
gyp ERR! command "/root/joy/manta-buckets-api/build/node/bin/node" "/root/joy/manta-buckets-api/build/node/lib/node_modules/npm/node_modules/node-gyp/bin/node-gyp.js" "rebuild"
gyp ERR! cwd /root/joy/manta-buckets-api/node_modules/bunyan/node_modules/dtrace-provider
gyp ERR! node -v v6.17.0
gyp ERR! node-gyp -v v3.4.0
gyp ERR! not ok
```

and for libmanta issues with node v6:

```
DTraceProviderBindings.target.mk:99: recipe for target 'Release/obj.target/DTraceProviderBindings/dtrace_provider.o' failed
make[1]: *** [Release/obj.target/DTraceProviderBindings/dtrace_provider.o] Error 1
make[1]: Leaving directory '/root/joy/manta-buckets-api/node_modules/libmanta/node_modules/restify/node_modules/dtrace-provider/build'
gyp ERR! build error
gyp ERR! stack Error: `make` failed with exit code: 2
gyp ERR! stack     at ChildProcess.onExit (/root/joy/manta-buckets-api/build/node/lib/node_modules/npm/node_modules/node-gyp/lib/build.js:276:23)
gyp ERR! stack     at emitTwo (events.js:106:13)
gyp ERR! stack     at ChildProcess.emit (events.js:191:7)
gyp ERR! stack     at Process.ChildProcess._handle.onexit (internal/child_process.js:219:12)
gyp ERR! System SunOS 5.11
gyp ERR! command "/root/joy/manta-buckets-api/build/node/bin/node" "/root/joy/manta-buckets-api/build/node/lib/node_modules/npm/node_modules/node-gyp/bin/node-gyp.js" "rebuild"
gyp ERR! cwd /root/joy/manta-buckets-api/node_modules/libmanta/node_modules/restify/node_modules/dtrace-provider
gyp ERR! node -v v6.17.0
gyp ERR! node-gyp -v v3.4.0
gyp ERR! not ok
```

I think these should probably be sorted first. The libmanta one might be a real
pain. I don't have much experience, but libmanta is a grabbag of tons of shared
crap. That it has its own *restify* is curious. Finding a shared restify version
that works for all users of libmanta might be "fun".

It is possible that cleaning libmanta up might be a part of the work here.
It is also possible that cleaning that up might be a lot easier once we have
the old mantav1 stuff branched off.

File: lib/buckets/common.js [hunk 5]
```diff
@@ -568,7 +568,14 @@ function paginationStream(opts, onRecord, done) {
...
 
-    log.debug('paginationStream starting');
+    log.debug({
+        vnodes: Object.keys(vnodes),
+        limit: limit,
+        order_by: opts.order_by,
+        delimiter: delimiter,
+        prefix: prefix
+    }, 'paginationStream starting');
+
```
Comment (trentm) [inline lib/buckets/common.js:578]: Some of these changes are not related to node v6 update right?

File: lib/common.js [hunk 6]
```diff
@@ -922,9 +921,9 @@ module.exports = {
             /*
              * MANTA-331: while a trailing '/' is ok in HTTP, this messes with
              * the consistent hashing, so ensure there isn't one
+             *
+             * This is now handled by the restify plugin sanitizePath
              */
-            /* JSSTYLED */
-            req._path = req._path.replace(/\/*$/, '');
```
Comment (trentm) [inline lib/common.js]: Perhaps just drop this comment and add a comment where sanitizePath() is used
mentioning its importance and why.

**Review by trentm — CHANGES_REQUESTED** [review]

File: package.json (hunk 4 above)
Comment (kellymclaughlin) [reply package.json:33]: Our usage of `libmanta` in `buckets-api` is pretty minimal and in fact I'd like to remove our usage of it altogether. There's just three functions we use and those could just be imported to the `buckets-api` `common` module.

File: lib/buckets/common.js [hunk 7]
```diff
@@ -74,7 +74,7 @@ function _list(type, req, bucket_id) {
         format('invalid _list type: %s', type));
     assert.object(req, 'req');
     assert.object(req.log, 'req.log');
-    assert.object(req.params, 'req.params');
+    assert.object(req.query, 'req.query');
     // assert.optionalUuid does not behave properly
```
Comment (kellymclaughlin) [inline lib/buckets/common.js:78]: I wonder if this behaves better with the newer version.

File: package.json (hunk 4 above)
Comment (trentm) [reply package.json:33]: +1 to dropping that dep then. It could very well be a node v6 mine field.  It really mostly feels (from my limited exposure) more like a "libmantadirstyle", if you know what I mean.

File: lib/buckets/common.js (hunk 7 above)
Comment (trentm) [reply lib/buckets/common.js:78]: Is there particular behaviour you mean?

I generally would advocate never using `req.params` in restify code: One should use `req.query` and/or `req.body` to be explicit.

Comment (kellymclaughlin) [reply lib/buckets/common.js:78]: I was referring to misbehavior of `assert.optionalUuid`. It the old version it was throwing asserts when a valid UUID was provided.

File: package.json (hunk 1 above)
Comment (bahamas10) [reply package.json:18]: will do

File: lib/buckets/common.js (hunk 7 above)
Comment (bahamas10) [reply lib/buckets/common.js:78]: I haven't bumped `assert-plus` as part of this PR.  I'm only trying to bump versions that are strictly required for node v6.  I want to do a follow-up PR after this one is merged where certain deps are updated (like assert-plus).

File: lib/buckets/common.js (hunk 5 above)
Comment (bahamas10) [reply lib/buckets/common.js:578]: Correct... but very little:

1. This log line - this was required when debugging req.params vs req.query changes from the restify upgrade, and I figured it was a good log line to keep in
2. configure.js also has the process.stderr stream change to fix the RequestCaptureStream which is also not related to node v6 AFAICT

File: lib/common.js (hunk 6 above)
Comment (bahamas10) [reply lib/common.js]: agreed

File: package.json (hunk 3 above)
Comment (bahamas10) [reply package.json:46]: Agreed, but I don't think that's needed for this PR.  I've created MANTA-4761 to track upgrading various deps in this repo (like verror and assert-plus)

File: package.json (hunk 2 above)
Comment (bahamas10) [reply package.json]: good catch.  I've updated errors to v5 to match what restify itself is using.

File: package.json (hunk 4 above)
Comment (bahamas10) [reply package.json:33]: I'm going to look into this today

File: lib/buckets/common.js (hunk 7 above)
Comment (kellymclaughlin) [reply lib/buckets/common.js:78]: Got it. Sorry, should have double-checked that before commenting.

File: package.json (hunk 4 above)
Comment (bahamas10) [reply package.json:33]: created MANTA-4764... if you both are cool with it, i can tackle this outside of the upgrade to node v6 since it's not currently blocking

File: lib/buckets/common.js (hunk 5 above)
Comment (trentm) [reply lib/buckets/common.js:578]: :+1:

File: package.json (hunk 4 above)
Comment (trentm) [reply package.json:33]: There are still "gyp: ERR!" build errors for:

- libmanta: I'm fine with defering that to MANTA-4764 as long as that doesn't linger too long
- watershed: Is there an updated watershed version with more modern dtrace-provider?
- smartdc: A devDependency. Does updating to smartdc@9.0.0 fix that? It looks like it might from this change:

```
commit db6910d25216e86658f82a80cfa1878664bf2721
Author: Brian Bennett
Date:   2015-10-22T21:23:25-07:00 (4 years, 1 month ago)

    Update dep to get newer dtrace-provider to fix build against node 4.x

diff --git a/package.json b/package.json
index 805ed18..05501c5 100644
--- a/package.json
+++ b/package.json
@@ -2,7 +2,7 @@
   "author": "Joyent, Inc. http://www.joyent.com",
   "name": "smartdc",
   "description": "Client SDK and CLI for the Joyent SmartDataCenter API",
-  "version": "7.5.3",
+  "version": "7.5.4",
```

Comment (bahamas10) [reply package.json:33]: 1. I'll take an MANTA-4764 and will get it done soon (before december)
2. good news! `watershed` isn't actually used... so let's remove it
3. i'll update to 9.0.0 and test it

```
$ rg watershed
package.json
48:        "watershed": "0.3.0",
```

^^ that's all that showed up

Comment (trentm) [reply package.json:33]: Great. 

FWIW, looks like the same for `strsplit` and `deep-equal` as well.  Checked via:

```
json -f package.json dependencies | json -Ma key \
    | while read dep; do echo ""; echo "# $dep"; rg $RG_OPTS --no-heading "require\('$dep['/]" </dev/null; done
json -f package.json devDependencies | json -Ma key \
    | while read dep; do echo ""; echo "# $dep"; rg $RG_OPTS --no-heading "require\('$dep['/]" </dev/null; done
```

Comment (bahamas10) [reply package.json:33]: @trentm good catch! i'm removing them now.  I'd love to see some form of that check in `make check` or something like that.  like `dependency "XXX" is unused`

**Review by trentm — APPROVED** [review]

File: package.json (hunk 4 above)
Comment (trentm) [reply package.json:33]: >  I'd love to see some form of that check in make check or something like that. like dependency "XXX" is unused

Might be hard to make that reliable. There are some deps that are valid for using their binaries (./node_modules/.bin/foo), but not `require`d.

**Review by kellymclaughlin — APPROVED** [review]
This is great! Thanks Dave!
